server/src/main/java/org/opensearch/indices/replication/checkpoint/RemoteStorePublishMergedSegmentAction.java

```diff
@@ -0,0 +1,217 @@
+import org.opensearch.cluster.action.shard.ShardStateAction;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.UploadListener;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.index.shard.IndexShard;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+import org.opensearch.indices.IndicesService;
+import org.opensearch.indices.replication.ActiveMergesRegistry;
+import org.opensearch.indices.replication.SegmentReplicationTargetService;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportService;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Replication action responsible for publishing merged segment to a remote store and
+ * notifying all replicas of newly available segments.
+ *
+ * @opensearch.api
+ */
+public class RemoteStorePublishMergedSegmentAction extends AbstractPublishCheckpointAction<
+    RemoteStorePublishMergedSegmentRequest,
+    RemoteStorePublishMergedSegmentRequest> implements MergedSegmentPublisher.PublishAction {
+
+    public static final String ACTION_NAME = "indices:admin/remote_publish_merged_segment";
+
+    private final static Logger logger = LogManager.getLogger(RemoteStorePublishMergedSegmentAction.class);
+
+    private final SegmentReplicationTargetService replicationService;
+
+    @Inject
+    public RemoteStorePublishMergedSegmentAction(
+        Settings settings,
+        TransportService transportService,
+        ClusterService clusterService,
+        IndicesService indicesService,
+        ThreadPool threadPool,
+        ShardStateAction shardStateAction,
+        ActionFilters actionFilters,
+        SegmentReplicationTargetService targetService
+    ) {
+        super(
+            settings,
+            ACTION_NAME,
+            transportService,
+            clusterService,
+            indicesService,
+            threadPool,
+            shardStateAction,
+            actionFilters,
+            RemoteStorePublishMergedSegmentRequest::new,
+            RemoteStorePublishMergedSegmentRequest::new,
+            ThreadPool.Names.GENERIC,
+            logger
+        );
+        this.replicationService = targetService;
+    }
+
+    @Override
+    protected void doReplicaOperation(RemoteStorePublishMergedSegmentRequest shardRequest, IndexShard replica) {
+        RemoteStoreMergedSegmentCheckpoint checkpoint = shardRequest.getMergedSegment();
+        if (checkpoint.getShardId().equals(replica.shardId())) {
+            if (Objects.equals(checkpoint.getPrimaryTerm(), replica.getOperationPrimaryTerm()) == false) {
```

Comment:
There are some changes done in https://github.com/opensearch-project/OpenSearch/pull/18720 which can be used probably to see if we are processing a newer checkpoint?